PhysicsPackages/IMCPhysicsPackage_class.f90

```diff
@@ -0,0 +1,507 @@
+  use IMCMaterial_inter,              only : IMCMaterial, IMCMaterial_CptrCast
+  use mgIMCMaterial_inter,            only : mgIMCMaterial
+
+  ! Operators
+  use collisionOperator_class,        only : collisionOperator
+  use transportOperator_inter,        only : transportOperator
+
+  ! Tallies
+  use tallyCodes
+  use tallyAdmin_class,               only : tallyAdmin
+  use tallyResult_class,              only : tallyResult
+  use imcWeightClerk_class,           only : imcWeightResult
+
+  ! Factories
+  use transportOperatorFactory_func,  only : new_transportOperator
+  use sourceFactory_func,             only : new_source
+
+  implicit none
+
+  private
+
+  !!
+  !! Physics Package for IMC calculations
+  !!
+  type, public,extends(physicsPackage) :: IMCPhysicsPackage
+    private
+    ! Building blocks
+    class(nuclearDatabase), pointer        :: nucData => null()
+    class(geometry), pointer               :: geom    => null()
+    integer(shortInt)                      :: geomIdx = 0
+    type(collisionOperator)                :: collOp
+    class(transportOperator), allocatable  :: transOp
+    class(RNG), pointer                    :: pRNG    => null()
+    type(tallyAdmin),pointer               :: tally   => null()
+    type(tallyAdmin),pointer               :: imcWeightAtch => null()
+
+    ! Settings
+    integer(shortInt)  :: N_cycles
+    integer(shortInt)  :: pop
+    integer(shortInt)  :: limit
+    real(defReal)      :: deltaT
+    character(pathLen) :: outputFile
+    character(nameLen) :: outputFormat
+    integer(shortInt)  :: printSource = 0
+    integer(shortInt)  :: particleType
+    logical(defBool)   :: sourceGiven = .false.
+    integer(shortInt)  :: nMat
+
+    ! Calculation components
+    type(particleDungeon), allocatable :: thisCycle
+    type(particleDungeon), allocatable :: nextCycle
+      ! Note that other physics packages used pointers for these particleDungeons ( => null() )
+      ! I found it easier to get 'allocatable' to work, unsure if this needs to be changed
+    class(source), allocatable     :: inputSource
+    class(source), allocatable     :: IMCSource
+
+    ! Timer bins
+    integer(shortInt)  :: timerMain
+    real (defReal)     :: CPU_time_start
+    real (defReal)     :: CPU_time_end
+
+  contains
+    procedure :: init
+    procedure :: printSettings
+    procedure :: cycles
+    procedure :: collectResults
+    procedure :: run
+    procedure :: kill
+
+  end type IMCPhysicsPackage
+
+contains
+
+  subroutine run(self)
+    class(IMCPhysicsPackage), intent(inout) :: self
+
+    print *, repeat("<>",50)
+    print *, "/\/\ IMC CALCULATION /\/\"
+
+    call self % cycles(self % tally, self % imcWeightAtch, self % N_cycles)
+    call self % collectResults()
+
+    print *
+    print *, "\/\/ END OF IMC CALCULATION \/\/"
+    print *
+  end subroutine
+
+  !!
+  !! Run cycles for calculation
+  !!
+  subroutine cycles(self, tally, tallyAtch, N_cycles)
```

Comment:
It's not major, but you can call this, say, steps instead if you like. We're not so attached to cycles when there's a physical meaning behind it.